### PR TITLE
ZAP integration with Owasp VulnerableApp

### PR DESCRIPTION
**Pull request by preetkaran20**
fixes https://github.com/zapbot/zap-mgmt-scripts/issues/31

### CODE REVIEW

File: .github/workflows/zap-vs-vulnerableapp.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+on:
+  workflow_dispatch:
+
+jobs:
+  scan:
+    runs-on: [ubuntu-latest]
+    steps:
+    - name: Start vulnerableApp
+      run: docker run --rm -d -p 9090:8080 --name=owasp-vulnerableapp sasanlabs/owasp-vulnerableapp:latest
```
Comment (preetkaran20) [inline .github/workflows/zap-vs-vulnerableapp.yml:11]: I am not sure how this script will actually run, created this file similar to wavsep.
@psiinon don't we need to first pull the vulnerableApp image ?

File: vulnerableApp/vulnerableApp_score.py [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script tests ZAP against Owasp VulnerableApp: https://github.com/SasanLabs/VulnerableApp
+# Note: Owasp VulnerableApp has to be installed somewhere - the above link is to the github
+# project not the test suite!
+#
+# To use this script:
+# * Install the ZAP Python API:
```
Comment (preetkaran20) [inline vulnerableApp/vulnerableApp_score.py:24]: faced issue because of Python2 in mac by default. do we need to consider adding the same in readme.md ?

**Comment by preetkaran20** [comment]
Sample report for reference:

[report.html.zip](https://github.com/zapbot/zap-mgmt-scripts/files/5079649/report.html.zip)

File: .github/workflows/zap-vs-vulnerableapp.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    - name: Start ZAP
+      run: docker run --rm -d -p 8090:8090 --network host owasp/zap2docker-weekly zap-x.sh -daemon -host localhost -port 8090 -addonupdate -addoninstall domxss -config "api.addrs.addr.name=.*" -config api.addrs.addr.regex=true -config api.disablekey=true &
```
Comment (preetkaran20) [inline .github/workflows/zap-vs-vulnerableapp.yml:13]: @psiinon i was looking to add the step of firing requests after spidering using the information provided by `scanner` endpoint for more better result comparison. Where can we do that ?

File: .github/workflows/zap-vs-vulnerableapp.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    - name: Run ZAP API script
+      env:
+        GITHUB_TOKEN: ${{ secrets.ZAPBOT_TOKEN }}
+      run: docker run --rm --network host --env ZAPBOT_TOKEN=${{ secrets.ZAPBOT_TOKEN }} owasp/zap-testing ./zap-vs-vulnerableApp.sh -e 32 -n vulnerableApp-weekly-RBJ-M-M -t "Rel, Beta, DomXSS"
```
Comment (preetkaran20) [inline .github/workflows/zap-vs-vulnerableapp.yml:17]: @psiinon i have updated files but as docker owasp/zap-testing is not having the files so the action is failing. can you please help in this.

File: dockerfile [hunk 5]
```diff
@@ -9,4 +9,4 @@ RUN pip3 install --upgrade python-owasp-zap-v2.4
 
 WORKDIR /zap
 
-copy wavsep wivet /zap/
+copy wavsep wivet vulnerableApp /zap/
```
Comment (preetkaran20) [inline dockerfile:12]: this is the change which is i think needed for the scope api to run @psiinon

Comment (preetkaran20) [reply dockerfile:12]: @psiinon i am getting following error while running this action: https://github.com/preetkaran20/zap-mgmt-scripts/runs/1054469186?check_suite_focus=true

can you please help with this.

Comment (preetkaran20) [reply dockerfile:12]: I think we need to first build the docker owasp/testing image and then we can try. How were you able to handle this.

File: .github/workflows/zap-vs-vulnerableapp.yml (hunk 4 above)
Comment (kingthorin) [reply .github/workflows/zap-vs-vulnerableapp.yml:17]: Pull the image?

Comment (kingthorin) [reply .github/workflows/zap-vs-vulnerableapp.yml:17]: Disregard for now, there are options but I’m camping and can’t dig in right now ⛺️

File: vulnerableApp/vulnerableApp_score.py [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+            s += '<a href="#' + i + '">' + i + '</a> '
+    return s
+
+
+def main(argv):
+    # -------------------------------------------------------------------------
+    # Default Configurations - use -h and -p for different host and port
+    # -------------------------------------------------------------------------
+    zapHost = '127.0.0.1'
+    zapPort = '8080'
```
Comment (psiinon) [inline vulnerableApp/vulnerableApp_score.py:57]: This causes a problem as the controlling script doesnt change it - ZAP is listening on 8090 and so this script cant connect..

**Comment by psiinon** [comment]
Looks like its failing when trying to score as its using 8080 for the ZAP port instead of 9090.
FYI to debug this I put in some debug statments and pushed those to a local branch.
The script used the owasp/zap-testing image so I recreated it with my debug statements and pushed it to psiinon/zap-testing
Example run: https://github.com/psiinon/zap-mgmt-scripts/runs/1209647711
You can see that the scoring script is trying to connect to ZAP via 8080.

**Comment by psiinon** [comment]
I've tried running it again with the port fixed but its now failing somewhere else: https://github.com/psiinon/zap-mgmt-scripts/runs/1209711657
@preetkaran20 is that enough to go on? When you test it yourself you'll need to use your own zap-testing docker image - you have to rebuld that if you make any changes to zap-vs-vulnerableApp.sh

**Review by psiinon — CHANGES_REQUESTED** [review]
ZAP port needs correcting and theres still another problem with the scoring

File: vulnerableApp/vulnerableApp_score.py (hunk 6 above)
Comment (preetkaran20) [reply vulnerableApp/vulnerableApp_score.py]: thanks alot for debugging this. Will look into it.

**Comment by preetkaran20** [comment]
> ZAP port needs correcting and theres still another problem with the scoring

Hi @psiinon ,

While running i am getting following error: https://github.com/preetkaran20/zap-mgmt-scripts/runs/1210105415?check_suite_focus=true and it looks like the Docker image is not having .sh file. do we need to create it in ZAP testing project ? 

thanks,
Karan

File: dockerfile (hunk 5 above)
Comment (psiinon) [reply dockerfile:12]: As per my previous comment :) I built it locally and pushed it to psiinon/zap-testing - I also changed my copy of the action to use that instead.

**Comment by preetkaran20** [comment]
@psiinon please find the glimpse of report: 
[Report.zip](https://github.com/zapbot/zap-mgmt-scripts/files/5470749/Report.zip)

**Comment by preetkaran20** [comment]
@psiinon I had ran the github action at: https://github.com/preetkaran20/zap-mgmt-scripts/runs/1337625905?check_suite_focus=true

File: .github/workflows/zap-vs-vulnerableapp.yml [hunk 7]
```diff
@@ -8,10 +8,10 @@ jobs:
     runs-on: [ubuntu-latest]
     steps:
     - name: Start vulnerableApp
-      run: docker run --rm -d -p 9090:8080 --name=owasp-vulnerableapp sasanlabs/owasp-vulnerableapp:latest
+      run: docker run --rm -d -p 9090:9090 --name=owasp-vulnerableapp sasanlabs/owasp-vulnerableapp:latest
```
Comment (preetkaran20) [inline .github/workflows/zap-vs-vulnerableapp.yml:11]: Problem earlier was this port.

File: vulnerableApp/vulnerableApp_score.py [hunk 8]
```diff
@@ -111,16 +152,14 @@ def main(argv):
     page = 1000
     # Page through the alerts as otherwise ZAP can hang...
     alerts = zap.core.alerts('', offset, page)
-
     # Scanner endpoint of Owasp VulnerableApp which provides information related to vulnerabilities present.
-    # TODO once vuln.yml specification is decided need to move this logic as per the spec
-    vulnerable_app_scanner_response = requests.get("http://127.0.0.1:9090/scanner");
+    vulnerable_app_scanner_response = requests.get("http://127.0.0.1:9090/scanner",
+                                                   proxies={'http': zapUrl, 'https': zapUrl}, verify=False)
```
Comment (preetkaran20) [inline vulnerableApp/vulnerableApp_score.py:157]: Another problem was not setting this proxy.

**Comment by psiinon** [comment]
So ZAP's not scoring very well then ;) Do you know why? I guess it will be a range of reasons.

Re the report:
The 'Group Scores' look like they should be at the top (and limit to 2 decimal places)
The 'Top Level Scores' dont look very top level - do they provide any value? 
The 'Detailed Results' look useful but only have the passes - if they included the failures then maybe we wouldnt need to 'Top Level Scores'?

**Comment by preetkaran20** [comment]
> So ZAP's not scoring very well then ;) Do you know why? I guess it will be a range of reasons.

I think the issue is only with Ajax Spider (as we discussed and that workaround is not working, may be a new version can do the trick), if Ajax Spider finds the correct URL/Query Param then the score is more than 80. 

> 
> Re the report:
> The 'Group Scores' look like they should be at the top (and limit to 2 decimal places)
group score should be at the top ? are you suggesting that group score should be first table ? I will make the decimal places as 2.
> The 'Top Level Scores' dont look very top level - do they provide any value?

So from my perspective, top level score is just telling, URL vs pass count/fail count/score (High level) but what has passed or failed or ignored or other is being shown by the detailed results. 
I think it is good to a one table that gives higher level view (count/score) and there is no need to go into too much details.
WDYT ?

> The 'Detailed Results' look useful but only have the passes - if they included the failures then maybe we wouldnt need to 'Top Level Scores'?

sure will add fail also to detailed results.

**Comment by preetkaran20** [comment]
Hi @psiinon 
I have done the changes as suggested. please review.
Glimpse of the report:
[ZAPVsVulnerableReport.zip](https://github.com/zapbot/zap-mgmt-scripts/files/5485377/ZAPVsVulnerableReport.zip)

thanks,
Karan

**Comment by preetkaran20** [comment]
[Uploading ZAP Vs VulnerableApp.zip…]()

New report.

**Review by psiinon — APPROVED** [review]

**Comment by preetkaran20** [comment]
@psiinon can we merge this PR ?

thanks,
Karan

**Pull request merged**